CreditManagement/models.py

```diff
@@ -1,196 +1,123 @@
+        return self.target_user if self.target_user else self.target_association
 
     def save(self, *args, **kwargs):
-        # If older versions are present, retract them
-        if (self.id):
-            # There is an older version get it.
-            old_version = Transaction.objects.get(id=self.id)
-            old_source = old_version.source()
-            old_target = old_version.target()
-
-            has_source_changed = old_source != self.source()
-            has_target_changed = old_target != self.target()
-
-            amount = self.amount - old_version.amount
-
-            with transaction.atomic():
-                if has_source_changed:
-                    if old_source is not None:
-                        credit = old_source.get_credit_containing_instance()
-                        credit.credit = F('credit') + old_version.amount
-                        credit.save()
-                    if self.source() is not None:
-                        credit = self.source().get_credit_containing_instance()
-                        credit.credit = F('credit') - self.amount
-                        credit.save()
-                elif amount != 0:
-                    if self.source() is not None:
-                        credit = self.source().get_credit_containing_instance()
-                        credit.credit = F('credit') - amount
-                        credit.save()
-
-                if has_target_changed:
-                    if old_target is not None:
-                        credit = old_target.get_credit_containing_instance()
-                        credit.credit = F('credit') - old_version.amount
-                        credit.save()
-                    if self.target() is not None:
-                        credit = self.target().get_credit_containing_instance()
-                        credit.credit = F('credit') + self.amount
-                        credit.save()
-                elif amount != 0:
-                    if self.target() is not None:
-                        credit = self.target().get_credit_containing_instance()
-                        credit.credit = F('credit') + amount
-                        credit.save()
-
-                super(Transaction, self).save(*args, **kwargs)
-
-        else:
-            with transaction.atomic():
-                if self.source() is not None:
-                    credit = self.source().get_credit_containing_instance()
-                    credit.credit = F('credit') - self.amount
-                    credit.save()
-                if self.target() is not None:
-                    credit = self.target().get_credit_containing_instance()
-                    credit.credit = F('credit') + self.amount
-                    credit.save()
-
-                super(Transaction, self).save(*args, **kwargs)
+        """
+        Double-checks database constraints.
+        """
+        if self.pk:
+            raise ValueError("Transaction change is not allowed")
```

Comment:
I wholeheartedly agree! Make some tests for it and turn these into checks into `assert`s

Reply:
I've put some of your suggestions in this commit: cd801be8bceab9826bad840cca1241238822c942, including test cases for the database constraints. 3 of them fail (the ones that do not check for `AssertionError`) because they check the constraints on the database level and these constraints are not yet into place.